File: java/CubicNoise.java
package noise;

public final class CubicNoise {
    private static final int RND_A = 134775813;
    private static final int RND_B = 1103515245;

    private int seed;
    private int octave;
    private int periodx = Integer.MAX_VALUE;
    private int periody = Integer.MAX_VALUE;

    public CubicNoise(final int seed, final int octave, final int periodx, final int periody)
    {
        this(seed, octave);

        this.periodx = periodx;
        this.periody = periody;
    }

    public CubicNoise(final int seed, final int octave)
    {
        this.seed = seed;
        this.octave = octave;
    }

    public float sample(final float x)
    {
        final int xi = (int) Math.floor(x / octave);
        final float lerp = x / octave - xi;
        
        return interpolate(
                randomize(seed, tile(xi - 1, periodx), 0),
                randomize(seed, tile(xi, periodx), 0),
                randomize(seed, tile(xi + 1, periodx), 0),
                randomize(seed, tile(xi + 2, periodx), 0),
                lerp) * 0.5f + 0.25f;
    }

    public float sample(final float x, final float y)
    {
        final int xi = (int) Math.floor(x / octave);
        final float lerpx = x / octave - xi;
        final int yi = (int) Math.floor(y / octave);
        final float lerpy = y / octave - yi;

        float xSamples[] = new float[4];

        for(int i = 0; i < 4; ++i)
            xSamples[i] = interpolate(
                    randomize(seed, tile(xi - 1, periodx), tile(yi - 1 + i, periody)),
                    randomize(seed, tile(xi, periodx), tile(yi - 1 + i, periody)),
                    randomize(seed, tile(xi + 1, periodx), tile(yi - 1 + i, periody)),
                    randomize(seed, tile(xi + 2, periodx), tile(yi - 1 + i, periody)),
                    lerpx);

        return interpolate(xSamples[0], xSamples[1], xSamples[2], xSamples[3], lerpy) * 0.5f + 0.25f;
    }

    private static float randomize(final int seed, int x, int y)
    {
        return (float) ((((x ^ y) * RND_A) ^ (seed + x)) * (((RND_B * x) << 16) ^ (RND_B * y) - RND_A)) / Integer.MAX_VALUE;
    }

    private static int tile(final int coordinate, final int period)
    {
        return coordinate % period;
    }

    private static float interpolate(final float a, final float b, final float c, final float d, final float x)
    {
        final float p = (d - c) - (a - b);

        return x * x * x * p + x * x * ((a - b) - p) + x * (c - a) + b;
    }
}

File: python/cubic_noise.py
from math import floor

# Feel free to change these
RND_A = 134775813
RND_B = 1103515245
SEED = 1
# Python3 integers are unlimited size.
# MAX_VAL is set so cubic_noise.py works the same as the other implementations.
MAX_VAL = 2**31 - 1
periodx = MAX_VAL
periody = MAX_VAL
# Don't change MASK. MASK is used for the bitwise operations in randomize()
MASK = 2**32 - 1

def interpolate(a, b, c, d, x):
    p = (d - c) - (a - b)
    return x * (x * (x * p + ((a - b) - p)) + (c - a)) + b

def randomize(x, y, seed=SEED):
    return (((((x ^ y) * RND_A) ^ (seed + x)) *
		(((RND_B * x) << 16) ^ (RND_B * y) - RND_A)) & MASK) / MASK

def tile(coordinate, period=MAX_VAL):
    return coordinate % period

def _sample1d(x, seed=SEED, octave=1):
    xi = floor(x / octave)
    lerp = x / octave - xi
    return interpolate(
                randomize(tile(xi - 1, periodx), 0, seed),
                randomize(tile(xi, periodx), 0, seed),
                randomize(tile(xi + 1, periodx), 0, seed),
                randomize(tile(xi + 2, periodx), 0, seed),
                lerp) * 0.5 + 0.25;

def _sample2d(x, y, seed=SEED, octave=1):
    xi = floor(x / octave)
    lerpx = x / octave - xi
    yi = floor(y / octave)
    lerpy = y / octave - yi
    
    xSamples = [0,0,0,0]
    
    for ii in range(4):
        xSamples[ii] = interpolate(
            randomize(tile(xi - 1, periodx), tile(yi - 1 + ii, periody), seed),
            randomize(tile(xi, periodx), tile(yi - 1 + ii, periody), seed),
            randomize(tile(xi + 1, periodx), tile(yi - 1 + ii, periody), seed),
            randomize(tile(xi + 2, periodx), tile(yi - 1 + ii, periody), seed),
            lerpx)
    
    return interpolate(*xSamples, lerpy) * 0.5 + 0.25;

def sample(x=None, y=None, seed=SEED, octave=1):
    if x is None:
        return _sample1d(y, seed, octave)
    elif y is None:
        return _sample1d(x, seed, octave)
    else:
        return _sample2d(x, y, seed, octave)
